ISPN-16730 Handle transactional context with many operations

jabolina opened the pull request:
https://issues.redhat.com/browse/ISPN-16730

The core class is the `TransactionDecorator` which creates a `DecoratedCache` to propagate the transactional context in all commands. It is really tricky because it needs to keep working the `EncoderCache` in any combination.

I also changed the functional map to utilize the methods in the DecoderCache to create the invocation context.

pruivo commented on server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java:
```diff
@@ -49,11 +48,18 @@ public class Resp3Handler extends Resp3AuthHandler {
+   }
+
+   @SuppressWarnings({"unchecked", "rawtypes"})
+   private <K, V> Cache<K, V> getCache(AdvancedCache cache) {
```
I'm not sure that I understand the use of this method. Can't `setCache` use `typedCache` directly?

jabolina replied:
Was double-checking this one. We need it to create the proper decoration. With this method, if the cache is DecoratedCache (after an EXEC), we call the `withMediaType` and create `DecoderCache` -> `DecoratedCache` -> `CacheImpl/SimpleCacheImpl` chain.

If we call just the `typedCache` method, it might return the `DecoratedCache` -> `CacheImpl/SimpleCacheImpl` chain.

I'll add this as a comment, otherwise, I forget what it does by tomorrow.

pruivo replied:
Where does the `DecoderCache` come from? I assume it is `EncoderCache` (there is the same typo in the comment).
What is the problem with `DecoratedCache` -> `[Simple]CacheImpl`? The `typedCache` method invokes only when it needs to re-encode the data as far as I understand. What am I missing?

jabolina replied:
Actually, I'll need to revert this whole block, it should always call the withMediaType before passing the instance to the multimap implementations. I'll add it as a comment, though.

Basically, the functional map implementation has a conversion in place (e.g., byte[] -> WrappedByteArray). If we don't pass an `EncoderCache` to these multimap implementations, they would try accessing entries with key (byte[]) instead for some operations. I thought we could use the `typedCache` to avoid that call in some cases, but we can not. Thanks for asking!

pruivo commented on server/resp/src/test/java/org/infinispan/server/resp/TransactionOperationsTest.java:
```diff
@@ -316,4 +323,55 @@ public void testAbortBecauseOfError() {
+   }
+
+   protected final void testBlockingPopWithTx(Supplier<String> keyFactory) throws Throwable {
+      boolean batch = cache.getCacheConfiguration().invocationBatching().enabled();
+      if (!batch)
+         throw new SkipException("Test does not have batching enabled! Skipping...");
+
```
nitpicking :) 
```suggestion
      SkipTestNG.skipIf(!cache.getCacheConfiguration().invocationBatching().enabled(), "Test does not have batching enabled.");
```

pruivo reviewed (commented):
LGTM but it requires an RESP expert to validate the tests :+1:

pruivo commented:
@jabolina BTW, it looks like batching is no longer required since we are creating the transactions manually. WDYT?

jabolina commented on commons/all/src/main/java/org/infinispan/commons/logging/Log.java:
```diff
@@ -297,6 +297,10 @@ public interface Log extends BasicLogger {
 
+   @LogMessage(level = WARN)
+   @Message(value = "Transactional cache should utilize PESSIMISTIC locking instead of '%s'", id = 971)
+   void utilizePessimisticLocking(String mode);
```
No, makes sense. It should be specific to RESP.

jabolina commented on commons/all/src/main/java/org/infinispan/commons/logging/Log.java:
```diff
@@ -297,6 +297,10 @@ public interface Log extends BasicLogger {
    RuntimeException certificateReloadError(@Cause Exception ex);
 
+   @LogMessage(level = WARN)
+   @Message(value = "Transactional cache should utilize PESSIMISTIC locking instead of '%s'", id = 971)
```
Pessimistic behaves more like Redis in the general case. I'll update the message.

jabolina commented:
Thanks! Updated with the suggestions.

> @jabolina BTW, it looks like batching is no longer required since we are creating the transactions manually. WDYT?

Yeah, makes sense. I've updated the checks and the log message.

wburns commented on server/resp/src/main/java/org/infinispan/server/resp/commands/tx/TransactionDecorator.java:
```diff
@@ -0,0 +1,175 @@
+      }
+
+      @Override
+      protected AdvancedCache<byte[], byte[]> withFlags(long newFlags) {
```
Should we also implement `noFlags`? Just to be safe.

wburns reviewed (commented):
LGTM, just a few comments.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/AbstractBlockingPop.java:
```diff
@@ -141,7 +141,19 @@ static CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCach
+    * </p>
+    *
+    * <p>
+    * <b>Warning:</b> Utilize an asynchronous listener to avoid deadlock in transactional caches. Otherwise, the listener
```
I think the issue is because the listener is called before the locks are released. So the poll operation tries to acquire the same lock the listener is holding. I didn't expect this as we use async operations. Also, this only happens in a TX cache.

wburns replied:
Yeah, it should be called before the locks are released. If the poll operation is acquiring a lock, we may want to offload that to the blocking thread pool then. Unfortunately, async listeners are invoked only on a single thread, to guarantee ordering between the same key, so it can be a bit of a scalability bottleneck.

jabolina replied:
Updated. I've removed the `sync=false` and instead offload the polling method to the non-blocking executor when receiving a notification. This is only for transactional caches.

wburns replied:
I am not sure what is going on here then. If the issue was it was deadlocking the invoking thread when it was sync and calling the poll directly. Then we cannot use the non blocking thread pool as that means it will block the non blocking thread until the lock is released as well, which is why I recommended the blocking pool.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/RespServer.java:
```diff
@@ -85,6 +86,10 @@ protected void startInternal() {
+            if (builder.transaction().lockingMode() != LockingMode.PESSIMISTIC)
+               org.infinispan.server.resp.logging.Log.CONFIG.utilizePessimisticLocking(builder.transaction().lockingMode().name());
+
          } else {
```
I was thinking about that, but I think it would affect perf for users not using MULTI/EXEC operations. We already have a warning message suggesting enabling TX. Otherwise, users would need to define a configuration to disable TX. WDYT?

wburns replied:
If we have a warning somewhere else, that is fine.

pruivo commented on server/resp/src/test/java/org/infinispan/server/resp/types/TransactionMediaTypeTest.java:
```diff
@@ -46,9 +48,9 @@ public Object[] factory() {
             MediaType.APPLICATION_OCTET_STREAM,
-            MediaType.APPLICATION_PROTOSTREAM,
+            /*MediaType.APPLICATION_PROTOSTREAM,
             MediaType.APPLICATION_OBJECT,
-            MediaType.TEXT_PLAIN,
+            MediaType.TEXT_PLAIN,*/
```
is this intentional?

jabolina replied:
Oops.

jabolina commented on server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/AbstractBlockingPop.java:
```diff
@@ -265,7 +278,11 @@ private void onEvent(byte[] key) {
          if (emptyQueue) {
-            runPoll(key);
+            if (executorService != null) {
+               CompletableFuture.runAsync(() -> runPoll(key), executorService);
```
We could revert all this offloading :). I've changed a check on the blocking manager that fixes the issue.

pruivo commented:
Optimistically merged! If anything is pending, open a follow-up PR @jabolina @wburns .
Thank you.
